### PR TITLE
HDFS-15479. Ordered snapshot deletion: make it a configurable feature

### PR DESCRIPTION
**Pull request by szetszwo**
Please see https://issues.apache.org/jira/browse/HDFS-15479

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  26m 11s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 32s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |   1m 24s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m  7s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 38s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 59s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 37s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m 41s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |   3m 39s |  hadoop-hdfs-project/hadoop-hdfs in trunk has 4 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 23s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 15s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |   1m 15s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 58s |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 4 new + 524 unchanged - 0 fixed = 528 total (was 524)  |
| +1 :green_heart: |  mvnsite  |   1m 26s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  18m 14s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 35s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 49s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   3m 59s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 114m 35s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   | 206m 25s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestDecommissioningStatus |
|   | hadoop.hdfs.server.datanode.TestDataNodeUUID |
|   | hadoop.hdfs.TestMaintenanceState |
|   | hadoop.hdfs.server.datanode.TestBPOfferService |
|   | hadoop.tools.TestHdfsConfigFields |
|   | hadoop.fs.contract.hdfs.TestHDFSContractMultipartUploader |
|   | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2156 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 3254a6045f44 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f2033de2342 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/testReport/ |
| Max. process+thread count | 3168 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 50s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  24m  7s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 31s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |   1m 15s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m  0s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 31s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  19m 47s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 37s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m 27s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |   3m 25s |  hadoop-hdfs-project/hadoop-hdfs in trunk has 4 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 27s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  9s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |   1m  9s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 54s |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 523 unchanged - 0 fixed = 524 total (was 523)  |
| +1 :green_heart: |  mvnsite  |   1m 16s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  16m 58s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 35s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 43s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   3m  4s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  98m  5s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 39s |  The patch does not generate ASF License warnings.  |
|  |   | 182m 49s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.fs.contract.hdfs.TestHDFSContractMultipartUploader |
|   | hadoop.tools.TestHdfsConfigFields |
|   | hadoop.hdfs.TestDFSInputStream |
|   | hadoop.hdfs.TestStripedFileAppend |
|   | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithDFSStriped |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2156 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux fb3c36b5aed7 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3833c616e08 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/testReport/ |
| Max. process+thread count | 4142 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 35s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  20m  4s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 13s |  trunk passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  compile  |   1m  8s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   0m 59s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 15s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 40s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 38s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |   2m 58s |  hadoop-hdfs-project/hadoop-hdfs in trunk has 4 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  7s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  8s |  the patch passed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04  |
| +1 :green_heart: |  javac  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  5s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  javac  |   1m  5s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 50s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 10s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  14m 13s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 34s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   3m 12s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 100m 17s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 42s |  The patch does not generate ASF License warnings.  |
|  |   | 172m 55s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |
|   | hadoop.fs.contract.hdfs.TestHDFSContractMultipartUploader |
|   | hadoop.tools.TestHdfsConfigFields |
|   | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithDFSStriped |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2156 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 378e6c9b5303 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3833c616e08 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.7+10-post-Ubuntu-2ubuntu218.04.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/testReport/ |
| Max. process+thread count | 3950 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2156/3/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by bshashikant** [comment]
Thanks @szetszwo for working on this. I have committed this.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java [hunk 1]
```diff
@@ -353,6 +354,20 @@ public int getListLimit() {
         + " hard limit " + DFSConfigKeys.DFS_NAMENODE_MAX_XATTR_SIZE_HARD_LIMIT
         + ": (%s).", DFSConfigKeys.DFS_NAMENODE_MAX_XATTR_SIZE_KEY);
 
+    this.snapshotDeletionOrdered =
+        conf.getBoolean(DFSConfigKeys.DFS_NAMENODE_SNAPSHOT_DELETION_ORDERED,
+            DFSConfigKeys.DFS_NAMENODE_SNAPSHOT_DELETION_ORDERED_DEFAULT);
+    LOG.info("{} = {}", DFSConfigKeys.DFS_NAMENODE_SNAPSHOT_DELETION_ORDERED,
+        snapshotDeletionOrdered);
+    if (snapshotDeletionOrdered && !xattrsEnabled) {
```
Comment (mukul1987) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java:362]: This check is not needed as all the other user of xattrs like encryption, erasure encoding calls
 FSDirXAttrOp.unprotectedSetXAttrs(fsd, srcIIP, xattrs, flag) directly without doing a feature check.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java [hunk 2]
```diff
@@ -500,8 +500,13 @@
 
   public static final String DFS_NAMENODE_SNAPSHOT_MAX_LIMIT =
       "dfs.namenode.snapshot.max.limit";
-
   public static final int DFS_NAMENODE_SNAPSHOT_MAX_LIMIT_DEFAULT = 65536;
+
+  public static final String DFS_NAMENODE_SNAPSHOT_DELETION_ORDERED =
+      "dfs.namenode.snapshot.deletion.ordered";
+  public static final boolean DFS_NAMENODE_SNAPSHOT_DELETION_ORDERED_DEFAULT
+      = false;
```
Comment (mukul1987) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java:508]: Lets also add the value to hdfs-default.xml

Comment (bshashikant) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java:508]: Addressed in https://issues.apache.org/jira/browse/HDFS-15480.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java (hunk 1 above)
Comment (bshashikant) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java:362]: https://issues.apache.org/jira/browse/HDFS-15480

**Pull request merged**